pkg/plugin/GlobalPluginService.go

```diff
@@ -1948,3 +1930,330 @@ func (impl *GlobalPluginServiceImpl) MigratePluginDataToParentPluginMetadata(plu
+	if !semver.IsValid(version) {
+		return &util.ApiError{
+			HttpStatusCode:  http.StatusBadRequest,
+			Code:            strconv.Itoa(http.StatusBadRequest),
+			InternalMessage: bean2.PluginVersionNotSemanticallyCorrectError,
+			UserMessage:     bean2.PluginVersionNotSemanticallyCorrectError,
+		}
+	}
+	//validate icon url and size
+	if len(pluginReq.Icon) > 0 {
+		err := utils.FetchIconAndCheckSize(pluginReq.Icon, bean2.PluginIconMaxSizeInBytes)
+		if err != nil {
+			return &util.ApiError{
+				HttpStatusCode:  http.StatusBadRequest,
+				Code:            strconv.Itoa(http.StatusBadRequest),
+				InternalMessage: fmt.Sprintf("%s err:= %s", bean2.PluginIconNotCorrectOrReachableError, err.Error()),
+				UserMessage:     fmt.Sprintf("%s err:= %s", bean2.PluginIconNotCorrectOrReachableError, err.Error()),
+			}
+		}
+	}
+	return nil
+}
+
+func (impl *GlobalPluginServiceImpl) createNewPlugin(tx *pg.Tx, pluginDto *bean2.PluginParentMetadataDto, userId int32) (int, error) {
+	pluginParentMetadata, err := impl.globalPluginRepository.SavePluginParentMetadata(tx, adaptor.GetPluginParentMetadataDbObject(pluginDto, userId))
+	if err != nil {
+		impl.logger.Errorw("createNewPlugin, error in saving plugin parent metadata", "pluginDto", pluginDto, "err", err)
+		return 0, err
+	}
+	pluginDto.Id = pluginParentMetadata.Id
+	pluginVersionDto := adaptor.GetPluginVersionMetadataDbObject(pluginDto, userId).
+		WithPluginParentMetadataId(pluginParentMetadata.Id).
+		WithIsLatestFlag(true)
+
+	pluginVersionMetadata, err := impl.globalPluginRepository.SavePluginMetadata(pluginVersionDto, tx)
+	if err != nil {
+		impl.logger.Errorw("createNewPlugin, error in saving plugin version metadata", "pluginDto", pluginDto, "err", err)
+		return 0, err
+	}
+	pluginDto.Versions.DetailedPluginVersionData[0].Id = pluginVersionMetadata.Id
+
+	pluginStageMapping := &repository.PluginStageMapping{
+		PluginId:  pluginParentMetadata.Id,
+		StageType: repository.CI_CD,
+		AuditLog:  sql.NewDefaultAuditLog(userId),
+	}
+	_, err = impl.globalPluginRepository.SavePluginStageMapping(pluginStageMapping, tx)
+	if err != nil {
+		impl.logger.Errorw("createNewPlugin, error in saving plugin stage mapping", "pluginDto", pluginDto, "err", err)
+		return 0, err
+	}
+
+	pluginStepsToCreate := pluginDto.Versions.DetailedPluginVersionData[0].PluginSteps
+	if len(pluginStepsToCreate) > 0 {
+		err = impl.saveDeepPluginStepData(pluginVersionMetadata.Id, pluginStepsToCreate, userId, tx)
+		if err != nil {
+			impl.logger.Errorw("createNewPlugin, error in saving plugin step data", "err", err)
+			return 0, err
+		}
+	} else {
+		return 0, util.NewApiError().WithCode(strconv.Itoa(http.StatusBadRequest)).WithHttpStatusCode(http.StatusBadRequest).
+			WithInternalMessage(bean2.PluginStepsNotProvidedError).WithUserMessage(errors.New(bean2.PluginStepsNotProvidedError))
+	}
+
+	err = impl.createPluginTagAndRelations(pluginDto.Versions.DetailedPluginVersionData[0], userId, tx)
```

Comment:
panic handling

Reply:
actually these are already handled in validations v2 func, which sits on top when service layer func in called